### PR TITLE
Milestone/open api integration

### PR DESCRIPTION
**Pull request by tomaszmichalak**
### Version used

* knot.x: 1.4.0

#### The current approach:
Knot.x defines two routes for processing requests:
* the templating engine that processes templates from repositories and injects data from external services
* Gateway API that allows you to create and expose REST API
This approach requires deep Knot.x knowledge to define a simple REST API interface. Additionally there is no validation mechanism for contracts.

#### Expected solution:
We want to use Open API 3 specification directly inside Knot.x code using design first approach. Additionally all request processors (handlers) should be configurable. Project specific processors should be easily configurable in routing configuration.

##### Example case:
The template repository authorizes requests based on the authentication header. This header comes from a cookie or (if it is not found) an external service (soft authentication / anonymous authentication). Thus, a custom authentication handler can be coded and configured before calling the repository.

The same authentication handler can be used in the Gateway API (for both custom flow and dedicated server).

### Appendix

* affects custom flow (Gateway API)

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -16,23 +16,42 @@ This is the repository for Knot.x core.
...
+### Microservices
+With [OpenAPI Specification (OAS)](https://github.com/OAI/OpenAPI-Specification/blob/master/versions/3.0.0.md) 
+support, you can define your RESTful APIs using standardized and language-agnostic approach. Your API 
+definition and its capabilities can be easily discovered and understood without access to source code or
+documentation. With the design first approach, your API consumer can interact with the remote service 
+with a minimal amount of implementation logic (all you need is to implement a simple 
+[`callback`](https://vertx.io/docs/apidocs/io/vertx/core/Handler.html) function, no Vert.x knowledge 
+is required). Knot.x combines [Vert.x-Web](https://vertx.io/docs/vertx-web/java/), 
+[Vert.x-Web API Contract](https://vertx.io/docs/vertx-web-api-contract/java/) and the 
+[Java ServiceLoader](https://docs.oracle.com/javase/7/docs/api/java/util/ServiceLoader.html) to make a
```
Comment (Skejven) [inline README.md]:
```suggestion
[Java ServiceLoader](https://docs.oracle.com/javase/8/docs/api/java/util/ServiceLoader.html) to make a
```

File: README.md [hunk 2]
```diff
@@ -16,23 +16,42 @@ This is the repository for Knot.x core.
...
+Knot.x allows to connect various systems like CMS, CRM and other in controlled and isolated way, 
+preventing any undesired interferences. It introduces an integration layer concept where Knot.x is an
+entry point that accepts all incoming requests, collects data from any datasource (like REST / SOAP 
+service, search engine, CRM etc.) and transforms it into an unified customer experience using a template from a
 [repository](https://github.com/Cognifide/knotx/wiki/RepositoryConnectors). The template can contain
 dynamic [fragments](https://github.com/Cognifide/knotx/wiki/Splitter) which determine the way how
-the data is used. The repository can be CMS system, Apache or simple catalogue from the filesystem with static HTML pages.
-
-For more information on Knot.x and where Knot.x core fits into the big picture please see the [KNOTX.IO](http://knotx.io).
+the data is used. All those steps require a minimal amount of implementation logic.
```
Comment (Skejven) [inline README.md]:
```suggestion
the data is used. All those steps require a minimal amount of implementation logic while custom implementations are easy to plug in.
```

File: knotx-core/src/main/java/io/knotx/server/KnotxServerVerticle.java [hunk 3]
```diff
@@ -55,95 +56,73 @@ public void init(Vertx vertx, Context context) {
...
-                .pathRegex(criteria.getPath())
-                .handler(KnotxAssemblerHandler.create(vertx, options));
-          }
-      );
+        );
+  }
+
+  private void registerHandlers(OpenAPI3RouterFactory routerFactory) {
+    List<RoutingHandlerFactory> handlerFactories = loadRoutingHandlerFactories();
+    options.getRoutingOperations().forEach(options -> {
```
Comment (Skejven) [inline knotx-core/src/main/java/io/knotx/server/KnotxServerVerticle.java]: Lambda param `options` covers `private KnotxServerOptions options` field. Please change it to:
```suggestion
    options.getRoutingOperations().forEach(operations -> {
    
```
And in lines below also change `options` to `operations`.

File: knotx-core/src/main/java/io/knotx/server/handler/http/request/body/BodyHandlerOptions.java [hunk 4]
```diff
@@ -0,0 +1,91 @@
...
+
+  public BodyHandlerOptions(JsonObject json) {
+    init();
+    if (json != null) {
+      BodyHandlerOptionsConverter.fromJson(json, this);
+    }
+  }
+
+  public void init() {
+    fileUploadLimit = DEFAULT_UPLOAD_LIMIT;
```
Comment (Skejven) [inline knotx-core/src/main/java/io/knotx/server/handler/http/request/body/BodyHandlerOptions.java]: You  may use value defined [here](https://github.com/vert-x3/vertx-web/blob/master/vertx-web/src/main/java/io/vertx/ext/web/handler/BodyHandler.java#L38) instead:
```suggestion
    fileUploadLimit = BodyHandler.DEFAULT_BODY_LIMIT;
```

File: knotx-core/src/main/java/io/knotx/server/handler/http/request/context/KnotContextRoutingHandlerFactory.java [hunk 5]
```diff
@@ -0,0 +1,43 @@
...
+  public String getName() {
+    return "knotContextHandler";
+  }
+
+  @Override
+  public Handler<RoutingContext> create(Vertx vertx, JsonObject config) {
+    return routingContext -> {
+      routingContext.put(KnotContext.KEY,
+          new KnotContext().setClientRequest(new ClientRequest(routingContext.request()))
+              .setClientResponse(new ClientResponse().setStatusCode(200).clearBody()));
```
Comment (Skejven) [inline knotx-core/src/main/java/io/knotx/server/handler/http/request/context/KnotContextRoutingHandlerFactory.java]:
```suggestion
              .setClientResponse(new ClientResponse().setStatusCode(HttpResponseStatus.OK.code()).clearBody()));
```

File: knotx-core/src/main/java/io/knotx/server/handler/http/response/cookie/CookieRoutingHandlerFactory.java [hunk 6]
```diff
@@ -1,5 +1,5 @@
 /*
- * Copyright (C) 2016 Cognifide Limited
+ * Copyright (C) 2018 Cognifide Limited
```
Comment (Skejven) [inline knotx-core/src/main/java/io/knotx/server/handler/http/response/cookie/CookieRoutingHandlerFactory.java]: `2016` was ok here :)

File: knotx-core/src/main/resources/io/knotx/logging/logback/base.xml [hunk 7]
```diff
@@ -24,7 +24,7 @@ Base logback configuration for knot.x
   <include resource="io/knotx/logging/logback/console-appender.xml"/>
   <include resource="io/knotx/logging/logback/file-appender.xml"/>
 
-  <root level="INFO">
+  <root level="TRACE">
```
Comment (Skejven) [inline knotx-core/src/main/resources/io/knotx/logging/logback/base.xml]: Let's stay with `INFO` by default.

File: knotx-core/src/test/resources/io/knotx/server/test-server.conf [hunk 8]
```diff
@@ -0,0 +1,134 @@
...
+              go-b {
+                address: B-engine,
+                onTransition.go-c.address: C-engine
+              },
+              go-c.address: C-engine
+            }
+          }
+        }
+      ] ${globals.routingOperations.handlers.response}
+      failureHandlers = [${globals.routingOperations.handlers.failure}]
```
Comment (Skejven) [inline knotx-core/src/test/resources/io/knotx/server/test-server.conf]: How about setting default failure handler in java code when none is defined here?

File: knotx-core/src/test/resources/logback-test.xml [hunk 9]
```diff
@@ -27,7 +27,9 @@
   </root>
 
   <logger name="io.netty" level="ERROR" additivity="false"/>
-  <logger name="io.vertx" level="ERROR" additivity="false"/>
+  <logger name="io.vertx" level="TRACE" additivity="false">
```
Comment (Skejven) [inline knotx-core/src/test/resources/logback-test.xml]: `ERROR`

File: knotx-core/src/test/java/io/knotx/junit/rule/TestVertxDeployer.java [hunk 10]
```diff
@@ -43,7 +43,7 @@ public Statement apply(final Statement base, Description description) {
       @Override
       public void evaluate() throws Throwable {
         KnotxConfiguration knotxConfig = description.getAnnotation(KnotxConfiguration.class);
-        if (knotxConfig == null || knotxConfig.value().isEmpty()) {
+        if (knotxConfig == null || knotxConfig.path().isEmpty() || knotxConfig.format().isEmpty()) {
           throw new IllegalArgumentException(
               "Missing @KnotxConfiguration annotation with the path to configuration JSON");
```
Comment (bmalecki) [inline knotx-core/src/test/java/io/knotx/junit/rule/TestVertxDeployer.java]: Configuration file format may be different than JSON

**Pull request merged**